File: autograder-web/src/api/submissions.ts
import { apiClient } from './client';

export interface Submission {
  id: number;
  exercise_id: number;
  student_id: number;
  code: string | null;
  status: 'queued' | 'running' | 'completed' | 'failed';
  submitted_at: string;
  error_message: string | null;
  file_name: string | null;
  file_size: number | null;
  content_type: string | null;
}

export interface SubmissionListItem {
  id: number;
  exercise_id: number;
  student_id: number;
  status: 'queued' | 'running' | 'completed' | 'failed';
  submitted_at: string;
  file_name: string | null;
}

export interface TestResultDetail {
  id: number;
  test_name: string;
  passed: boolean;
  message: string | null;
  stdout: string | null;
  stderr: string | null;
}

export interface LLMEvaluation {
  id: number;
  feedback: string;
  score: number;
  cached: boolean;
  created_at: string;
}

export interface GradeDetail {
  id: number;
  test_score: number | null;
  llm_score: number | null;
  final_score: number;
  late_penalty_applied: number;
  published: boolean;
}

export interface RubricScoreDetail {
  dimension_name: string;
  dimension_weight: number;
  score: number;
  feedback: string | null;
}

export interface SubmissionDetail {
  submission: Submission;
  test_results: TestResultDetail[] | null;
  llm_evaluation: LLMEvaluation | null;
  grade: GradeDetail | null;
  rubric_scores: RubricScoreDetail[] | null;
  overall_feedback: string | null;
}

export const submissionsApi = {
  list: async (params?: { exercise_id?: number; student_id?: number }) => {
    const { data } = await apiClient.get<SubmissionListItem[]>('/submissions', { params });
    return data;
  },

  get: async (id: number) => {
    const { data } = await apiClient.get<Submission>(`/submissions/${id}`);
    return data;
  },

  getResults: async (id: number) => {
    const { data } = await apiClient.get<SubmissionDetail>(`/submissions/${id}/results`);
    return data;
  },

  getStatus: async (id: number) => {
    const { data } = await apiClient.get<{ id: number; status: string; error_message: string | null }>(
      `/submissions/${id}/status`
    );
    return data;
  },

  submit: async (exerciseId: number, code: string) => {
    const formData = new FormData();
    formData.append('exercise_id', String(exerciseId));
    formData.append('code', code);
    const { data } = await apiClient.post<Submission>('/submissions', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return data;
  },

  submitFile: async (exerciseId: number, file: File) => {
    const formData = new FormData();
    formData.append('exercise_id', String(exerciseId));
    formData.append('file', file);
    const { data } = await apiClient.post<Submission>('/submissions', formData, {
      headers: { 'Content-Type': 'multipart/form-data' },
    });
    return data;
  },

  getDiff: async (submissionId: number, comparisonId: number) => {
    const { data } = await apiClient.get<string>(
      `/submissions/${submissionId}/diff/${comparisonId}`
    );
    return data;
  },
};

File: autograder-back/app/schemas/submissions.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum


class SubmissionStatus(str, Enum):
    """Submission processing status"""
    QUEUED = "queued"
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class SubmissionCreate(BaseModel):
    """Schema for creating a submission via text input"""
    exercise_id: int = Field(..., gt=0)
    code: Optional[str] = Field(None, min_length=1)


class SubmissionResponse(BaseModel):
    """Schema for submission response"""
    id: int
    exercise_id: int
    student_id: int
    code: Optional[str] = None
    status: SubmissionStatus
    submitted_at: datetime
    error_message: Optional[str] = None

    # File upload fields
    file_name: Optional[str] = None
    file_size: Optional[int] = None
    content_type: Optional[str] = None

    class Config:
        from_attributes = True


class SubmissionListResponse(BaseModel):
    """Schema for listing submissions"""
    id: int
    exercise_id: int
    student_id: int
    status: SubmissionStatus
    submitted_at: datetime
    file_name: Optional[str] = None

    class Config:
        from_attributes = True


class TestResultResponse(BaseModel):
    """Schema for test result response"""
    id: int
    test_name: str
    passed: bool
    message: Optional[str] = None
    stdout: Optional[str] = None
    stderr: Optional[str] = None

    class Config:
        from_attributes = True


class LLMEvaluationResponse(BaseModel):
    """Schema for LLM evaluation response"""
    id: int
    feedback: str
    score: float
    cached: bool
    created_at: datetime

    class Config:
        from_attributes = True


class GradeResponse(BaseModel):
    """Schema for grade response"""
    id: int
    test_score: Optional[float] = None
    llm_score: Optional[float] = None
    final_score: float
    late_penalty_applied: float
    published: bool

    class Config:
        from_attributes = True


class RubricScoreResponse(BaseModel):
    """Schema for rubric score response"""
    dimension_name: str
    dimension_weight: float
    score: float
    feedback: Optional[str] = None

    class Config:
        from_attributes = True


class SubmissionDetailResponse(BaseModel):
    """Schema for detailed submission response with results"""
    submission: SubmissionResponse
    test_results: Optional[List[TestResultResponse]] = None
    llm_evaluation: Optional[LLMEvaluationResponse] = None
    grade: Optional[GradeResponse] = None
    rubric_scores: Optional[List[RubricScoreResponse]] = None
    overall_feedback: Optional[str] = None
